compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/ir/ScopedValueMappingTransformer.kt

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright 2010-2019 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.backend.jvm.ir
+
+import org.jetbrains.kotlin.backend.common.pop
+import org.jetbrains.kotlin.ir.declarations.IrValueDeclaration
+import org.jetbrains.kotlin.ir.expressions.IrExpression
+import org.jetbrains.kotlin.ir.expressions.IrGetValue
+import org.jetbrains.kotlin.ir.expressions.IrSetVariable
+import org.jetbrains.kotlin.ir.expressions.impl.IrGetValueImpl
+import org.jetbrains.kotlin.ir.expressions.impl.IrSetVariableImpl
+import org.jetbrains.kotlin.ir.symbols.IrValueSymbol
+import org.jetbrains.kotlin.ir.symbols.IrVariableSymbol
+import org.jetbrains.kotlin.ir.visitors.IrElementTransformerVoid
+
+/**
+ * Replaces value parameters in GetValue/SetVariable based on a hierarchical mapping between
+ * value parameters.
+ *
+ * This is useful if you copy (part of) a function body to a new function. In this case,
+ * we want to update the value parameters in the scope of the new function, without updating
+ * the parameters in the scope of the existing function. Using this class allows us to avoid
+ * the quadratic overhead of nested calls to, e.g., deepCopyWithSymbols.
+ */
+open class ScopedValueMappingTransformer : IrElementTransformerVoid() {
+    protected val valueMap = ScopedValueDeclarationMap()
+
+    protected inline fun <T> scoped(block: () -> T): T {
+        valueMap.push()
+        return try {
+            block()
+        } finally {
+            valueMap.pop()
+        }
+    }
+
+    protected fun addMapping(symbol: IrValueSymbol, declaration: IrValueDeclaration) {
+        valueMap[symbol] = declaration
+    }
+
+    protected fun addMappings(mappings: Map<out IrValueSymbol, IrValueDeclaration>) =
+        valueMap.addAll(mappings)
+
+    override fun visitGetValue(expression: IrGetValue): IrExpression {
+        valueMap[expression.symbol]?.let {
+            return IrGetValueImpl(
+                expression.startOffset, expression.endOffset,
+                it.type, it.symbol, expression.origin
+            )
+        }
+        return super.visitGetValue(expression)
+    }
+
+    override fun visitSetVariable(expression: IrSetVariable): IrExpression {
+        valueMap[expression.symbol]?.let {
+            return IrSetVariableImpl(
+                expression.startOffset, expression.endOffset,
+                it.type, it.symbol as IrVariableSymbol, expression.origin
+            ).apply {
+                value = expression.value.transform(this@ScopedValueMappingTransformer, null)
+            }
+        }
+        return super.visitSetVariable(expression)
+    }
+}
+
+class ScopedValueDeclarationMap {
+    private val scopeStack = mutableListOf<MutableMap<IrValueSymbol, IrValueDeclaration>>(mutableMapOf())
```

Comment:
Could you please explain (with sample code maybe) why is a stack needed here and why is it not just a big mutable map? I got an impression that keys (`IrValueSymbol` instances) are unique here, which could allow us to simplify code by not bothering to remove entries, but I think I'm missing something

Reply:
Looking at this code again, I don't think it's necessary anymore. I've removed it and simplified the code.

For completeness though, the use case is explained in the comment to `ScopedValueMappingTransformer`. It's for the situation where you copy parts of one function into another. In this case you want to remap the value parameters in the replacement function but not in the body of the original, yet you have to traverse both.

Previously, the lowering also added coercions and changed types. That was where the second traversal came from. Right now the lowering is simpler so that a map suffices. I've changed it and removed the transformer.